File: go_app/src/go/src/pkg/sandc/sandcproxy.go
package sandc 
import (
	"fmt"
	"errors"
	"time"
	"encoding/json" 
)

type ProxyMsgCb func(ins *SandCProxy, topic, payload string) 

type SandCProxy struct {
	seq int32
	sc *SandC
	srv_topics []string
	cli_topics []string
	on_message ProxyMsgCb
	out_seq_map map[int32] chan interface{}
}

type ProxyRequest struct {
	Mod string 						`json:"mod"`
	Seq int32  						`json:"seq"`	
	Pld map[string] interface{} 	`json:"pld"`
}

func NewSandCProxy(clientid string) *SandCProxy {
	sc := NewSandC(clientid)
	ins := &SandCProxy{
		sc:sc,
		seq:1,
		on_message:func(psc *SandCProxy, topic, payload string) {},
	}
	ins.out_seq_map = make(map[int32]chan interface{})

	(*sc).SetMsgCallback(func(ig *SandC, m map[string]string) {
		tp, ok := m["tp"]
		if ok == false {
			return
		}

		pl, ok := m["pl"]
		if ok == false {
			return
		}
		
		is_client_topic := false 
		for _, topic := range (*ins).cli_topics {
			if topic == tp {
				is_client_topic = true
				break
			}
		}

		if is_client_topic == false {
			(*ins).on_message(ins, tp, pl)
			return
		}

		var p ProxyRequest
		err := json.Unmarshal([]byte(pl), &p)
		if err != nil {
			panic(err)
		}

		seq := p.Seq 	
		if seq <= 0 {
			panic("invalid seq " + string(seq))
		}
		
		ch, ok := ins.out_seq_map[seq]
		if ok == false {
			fmt.Println("missing seq", seq)
		} else {
			ch <- p.Pld
			close(ch)
			delete(ins.out_seq_map, seq)
		}
	})
	return ins
}

func (ins *SandCProxy) Run() {
	(*ins).sc.Run()
}

func (ins *SandCProxy) Query(topic string, data map[string] interface{}, timeout time.Duration) (interface{}, error) {
	seq := (*ins).seq;
	(*ins).seq = (*ins).seq + 1

	r := ProxyRequest {
		Mod : (*ins).cli_topics[0],
		Seq : seq,
		Pld : data,
	}

	b, err := json.Marshal(r)
	if err != nil {
		panic("Marshal fail")
	}

	ch := make(chan interface{})
	(*ins).out_seq_map[seq] = ch

	var res interface{}
	err = (*ins).sc.Publish(topic, string(b))
	if err != nil {
		return res, err
	}

	select {
	case res = <- ch:
	case <- time.After(timeout*time.Second):
		close(ch)
		err = errors.New("timeout")
		delete((*ins).out_seq_map, seq)
	}

	if len((*ins).out_seq_map) > 20 {
		panic("too many wait " + string(len((*ins).out_seq_map)))
	}

	return res, err 
}

func (ins *SandCProxy) SetMsgCallback(f ProxyMsgCb) {
	(*ins).on_message = f
}

func (ins *SandCProxy) Publish(topic, payload string) error {
	return (*ins).sc.Publish(topic, payload)
}

func (ins *SandCProxy) PreSubscribe(client_topics, server_topics []string) {
	(*ins).srv_topics = server_topics
	(*ins).cli_topics = client_topics
	tmp := client_topics
	for _, v := range server_topics {
		tmp = append(tmp, v)
	}
	(*ins).sc.PreSubscribe(tmp)
}

func (ins *SandCProxy) SetKeepalive(kp int64) {
	(*ins).sc.SetKeepalive(kp)
}

func (ins *SandCProxy) SetConnect(topic, payload string) {
	(*ins).sc.SetConnect(topic, payload)
}

func (ins *SandCProxy) SetWill(topic, payload string) {
	(*ins).sc.SetWill(topic, payload)
}

func (ins *SandCProxy) Connect(addr string) error {
	return (*ins).sc.Connect(addr)
}

func (ins *SandCProxy) SetAuth(username,  password string) {
	(*ins).sc.SetAuth(username, password)
}

File: sands/files/lib/sandcproxy.lua
local ski = require("ski") 
local sandc = require("sandc")
local js = require("cjson.safe") 

local function numb()  end

local method = {}
local mt = {__index = method}

function method:running() return self.mqtt:running() end
function method:set_callback(name, cb) self.mqtt:set_callback(name, cb) end
function method:publish(...) return self.mqtt:publish(...) end
function method:disconnect() return self.mqtt:disconnect() end
function method:connect(...) return self.mqtt:connect(...) end 
function method:set_auth(...) return self.mqtt:set_auth(...) end
function method:set_will(...) return self.mqtt:set_will(...) end
function method:set_connect(...) return self.mqtt:set_connect(...) end
function method:set_keepalive(...) return self.mqtt:set_keepalive(...) end
function method:set_message_cb(cb) self.on_message = cb end 
function method:set_proxy_topic(t) self.proxy_topic = t end
function method:run() 
	local timeout_item = {nil, "timeout"}
	ski.go(function()
		while self.mqtt:running() do 
			ski.sleep(1)
			local now, del, out_seq_map = ski.time(), {}, self.out_seq_map
			for seq, v in pairs(out_seq_map) do
				local _ = now - v.t > v.m and table.insert(del, seq)
			end
			for _, seq in ipairs(del) do
				out_seq_map[seq].ch:write(timeout_item) 
			end
		end
	end)
	return self.mqtt:run() 
end 

function method:pre_subscribe(client_topics, server_topics)
	local arr = {}
	for _, tp in ipairs(client_topics) do 
		self.cli_topics[tp] = 1, table.insert(arr, tp)
		self.cli_topic = self.cli_topic and self.cli_topic or tp
	end
	for _, tp in ipairs(server_topics or {}) do 
		self.srv_topics[tp] = 1, table.insert(arr, tp)
	end

	assert(self.cli_topic)
	return self.mqtt:pre_subscribe(unpack(arr))  
end

function method:query(topic, data, timeout)
	local seq, ch = self.seq, ski.new_chan(1)
	self.seq = self.seq + 1 

	local r, seq_item = self.query_item, {}
	r.seq, r.pld, r.mod = seq, data, self.cli_topic	
	seq_item.ch, seq_item.m, seq_item.t = ch, timeout or 3, ski.time()
	self.out_seq_map[seq] = seq_item
	local s = js.encode(r)
	local ret, err = self.mqtt:publish(topic, s)	 	assert(ret, err)
	local r, e = ch:read()		 						assert(r, e)
	self.out_seq_map[seq] = nil
	ch:close()
	return r[1], r[2]
end

function method:query_r(topic, data, timeout)
	local seq, ch = self.seq, ski.new_chan(1)
	self.seq = self.seq + 1 
	local r, seq_item = {out_topic = topic, data = {mod = self.cli_topic, seq = seq, pld = data}}, {}
	seq_item.ch, seq_item.m, seq_item.t = ch, timeout or 3, ski.time()
	self.out_seq_map[seq] = seq_item
	local s = js.encode(r) 
	local ret, err = self.mqtt:publish(self.proxy_topic, s)	 	assert(ret, err)
	local r, e = ch:read()		 								assert(r, e)
	self.out_seq_map[seq] = nil
	ch:close()
	return r[1], r[2]
end

function method:publish_r(topic, data)   
	local s = js.encode({out_topic = topic, data = {pld = data}})
	return self.mqtt:publish(self.proxy_topic, s)
end

local function new(clientid)
	local mqtt = sandc.new(clientid) 

	local obj = {
		seq = 0, 
		mqtt = mqtt, 
		cli_topic = nil,
		cli_topics = {},
		srv_topics = {},
		out_seq_map = {},  
		on_message = numb, 
		query_item = {},
		proxy_topic = "a/ac/proxy",
	}

	local content = {}
	mqtt:set_callback("on_message", function(topic, payload)
		if topic ~= obj.cli_topic then
			return obj.on_message(topic, payload)
		end

		local map = js.decode(payload)
		if not (map and map.seq and map.pld) then 
			return obj.on_message(topic, payload)
		end

		local seq = math.floor(tonumber(map.seq))
		local n = obj.out_seq_map[seq]
		if n then
			content[1] = map.pld
			n.ch:write(content)
		end
	end)

	setmetatable(obj, mt)
	return obj
end

local function run_new(map)
	assert(map and map.on_message and map.on_disconnect and map.unique and map.log)
	local proxy = new(map.unique)
	proxy:set_auth("ewrdcv34!@@@zvdasfFD*s34!@@@fadefsasfvadsfewa123$", "1fff89167~!223423@$$%^^&&&*&*}{}|/.,/.,.,<>?")
	proxy:pre_subscribe(map.clitopic or {}, map.srvtopic or {})
	proxy:set_message_cb(map.on_message) 
	local _ = map.on_disconnect and proxy:set_callback("on_disconnect", map.on_disconnect)
	local host, port = "127.0.0.1", 61886
	local ret, err = proxy:connect(map.host or "127.0.0.1", map.port or 61886)
	local _ = ret or map.log.fatal("connect fail %s", err)
	proxy:run()
	return proxy 
end

return {new = new, run_new = run_new}

